lib/validator.js

```diff
@@ -502,6 +502,72 @@ class Validator {
 		if (typeof fn !== "function") throw new Error("Plugin fn type must be function");
 		return fn(this);
 	}
+
+	/**
+	 * Normalize a schema, type or short hand definition by expanding it to a full form. The 'normalized'
+	 * form is the equivalent schema with any short hands undone. This ensure that each rule; always includes
+	 * a 'type' key, arrays always have an 'items' key, 'multi' always have a 'rules' key and objects always
+	 * have their properties defined in a 'props' key
+	 *
+	 * @param { * } value The value to normalize
+	 * @returns {Object} The normalized form of the given rule or schema
+	 */
+	normalize(value) {
+		let result;
+
+		if(typeof value === "string")
+			result = deepExtend({}, this.parseShortHand(value));
+		if(typeof result === "undefined")
+			result = deepExtend({}, value);
+		if(this.aliases[result.type])
+			result = deepExtend(result, this.normalize(this.aliases[result.type]), { skipIfExists: true});
+
+		result = deepExtend(result, this.defaults[result.type], { skipIfExist: true });
+
+		if(result.$$type) { // assume object type with other config
+			const type = result.$$type;
+			delete result.$$type;
+			const props = deepExtend({}, result);
+			result = this.normalize(type);
+			result.props = this.normalize(props);
+			return result;
+		}
+		if(Array.isArray(value)) {
+			result = {
+				type: "multi",
+				rules: value
+			};
+		}
+		if(result.type === "multi") {
+			result.rules = result.rules.map(r => this.normalize(r));
+			result.optional = result.rules.every(r => r.optional === true);
+			return result;
+		}
+		if(result.type === "array") {
+			result.items = this.normalize(result.items);
+			return result;
+		}
+		if(result.type === "object") {
+			if(result.props) {
+				for (const [k, v] of Object.entries(result.props)) {
+					result.props[k] = this.normalize(v);
+				}
+			}
+		}
+		if(typeof value === "object") {
+			if(value.type) {
+				const config = this.normalize(value.type);
+				deepExtend(result, config, { skipIfExists: true });
+			}
+			else{
+				for (const [k, v] of Object.entries(value)) {
```

Comment:
@icebob This is an ES6 syntax, should we done #240 sooner?

Reply:
It should work with rollup as well, buble should convert it to ES5.